File: sdks/nodejs/src/compute.ts
import { CloudPoolClient } from './index';

export interface CreateDeploymentConfig {
  name: string;
  image: string;
  replicas?: number;
  env?: Record<string, string>;
  ports?: number[];
  resources?: { cpu?: string; memory?: string };
}

export interface CreateCronJobConfig {
  name: string;
  schedule: string;
  command: string;
  image?: string;
  env?: Record<string, string>;
}

export interface DeployServerlessConfig {
  name: string;
  runtime: string;
  source: string;
  entrypoint?: string;
  env?: Record<string, string>;
  memory?: string;
  timeout?: number;
}

export class ComputeClient {
  constructor(private readonly client: CloudPoolClient) {}

  public async listDeployments<T = unknown>(params?: {
    projectId?: string;
  }): Promise<T[]> {
    const query = params?.projectId
      ? `?projectId=${encodeURIComponent(params.projectId)}`
      : '';
    const response = await this.client.request(
      'GET',
      `/api/v1/compute/deployments${query}`
    );
    return response.json();
  }

  public async createDeployment<T = unknown>(
    config: CreateDeploymentConfig
  ): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/compute/deployments',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config),
      }
    );
    return response.json();
  }

  public async deleteDeployment<T = unknown>(id: string): Promise<void> {
    await this.client.request(
      'DELETE',
      `/api/v1/compute/deployments/${encodeURIComponent(id)}`
    );
  }

  public async listCronJobs<T = unknown>(params?: {
    projectId?: string;
  }): Promise<T[]> {
    const query = params?.projectId
      ? `?projectId=${encodeURIComponent(params.projectId)}`
      : '';
    const response = await this.client.request(
      'GET',
      `/api/v1/compute/cron-jobs${query}`
    );
    return response.json();
  }

  public async createCronJob<T = unknown>(config: CreateCronJobConfig): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/compute/cron-jobs',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config),
      }
    );
    return response.json();
  }

  public async deleteCronJob<T = unknown>(id: string): Promise<void> {
    await this.client.request(
      'DELETE',
      `/api/v1/compute/cron-jobs/${encodeURIComponent(id)}`
    );
  }

  public async listServerless<T = unknown>(params?: {
    projectId?: string;
  }): Promise<T[]> {
    const query = params?.projectId
      ? `?projectId=${encodeURIComponent(params.projectId)}`
      : '';
    const response = await this.client.request(
      'GET',
      `/api/v1/compute/serverless${query}`
    );
    return response.json();
  }

  public async deployServerless<T = unknown>(
    config: DeployServerlessConfig
  ): Promise<T> {
    const response = await this.client.request(
      'POST',
      '/api/v1/compute/serverless',
      {
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(config),
      }
    );
    return response.json();
  }

  public async listLogs<T = unknown>(params?: {
    resourceId?: string;
    resourceType?: string;
    limit?: number;
    since?: string;
  }): Promise<T[]> {
    const searchParams = new URLSearchParams();
    if (params?.resourceId)
      searchParams.set('resourceId', params.resourceId);
    if (params?.resourceType)
      searchParams.set('resourceType', params.resourceType);
    if (params?.limit) searchParams.set('limit', String(params.limit));
    if (params?.since) searchParams.set('since', params.since);
    const query = searchParams.toString();
    const response = await this.client.request(
      'GET',
      `/api/v1/compute/logs${query ? '?' + query : ''}`
    );
    return response.json();
  }

  public async listPods<T = unknown>(params?: {
    deploymentId?: string;
  }): Promise<T[]> {
    const query = params?.deploymentId
      ? `?deploymentId=${encodeURIComponent(params.deploymentId)}`
      : '';
    const response = await this.client.request(
      'GET',
      `/api/v1/compute/pods${query}`
    );
    return response.json();
  }
}

File: sdks/rust/src/compute.rs
use reqwest::Method;
use serde_json::Value;

use crate::client::CloudPoolClient;
use crate::errors::CloudPoolError;

/// Client for managing compute deployments, cron jobs, serverless functions, logs, and pods.
pub struct ComputeClient<'a> {
    client: &'a CloudPoolClient,
}

impl<'a> ComputeClient<'a> {
    /// Create a new compute client.
    pub fn new(client: &'a CloudPoolClient) -> Self {
        Self { client }
    }

    /// List all compute deployments.
    pub async fn list_deployments(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "compute/deployments"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Create a new compute deployment.
    pub async fn create_deployment(&self, config: Value) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::POST, "compute/deployments").json(&config))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Delete a compute deployment by ID.
    pub async fn delete_deployment(&self, deployment_id: &str) -> Result<Value, CloudPoolError> {
        let path = format!("compute/deployments/{}", deployment_id);
        let res = self
            .client
            .send_with_retry(self.client.request(Method::DELETE, &path))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List all cron jobs.
    pub async fn list_cron_jobs(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "compute/cron-jobs"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Create a new cron job.
    pub async fn create_cron_job(&self, config: Value) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::POST, "compute/cron-jobs").json(&config))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Delete a cron job by ID.
    pub async fn delete_cron_job(&self, job_id: &str) -> Result<Value, CloudPoolError> {
        let path = format!("compute/cron-jobs/{}", job_id);
        let res = self
            .client
            .send_with_retry(self.client.request(Method::DELETE, &path))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List all serverless function deployments.
    pub async fn list_serverless(&self) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::GET, "compute/serverless"))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// Deploy a new serverless function.
    pub async fn deploy_serverless(&self, config: Value) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(self.client.request(Method::POST, "compute/serverless").json(&config))
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List logs for a specified resource.
    pub async fn list_logs(&self, resource_id: &str) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(
                self.client
                    .request(Method::GET, "compute/logs")
                    .query(&[("resourceId", resource_id)]),
            )
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }

    /// List pods for a given deployment or resource.
    pub async fn list_pods(&self, deployment_id: &str) -> Result<Value, CloudPoolError> {
        let res = self
            .client
            .send_with_retry(
                self.client
                    .request(Method::GET, "compute/pods")
                    .query(&[("deploymentId", deployment_id)]),
            )
            .await?;
        Ok(res.json::<Value>().await.map_err(|e| CloudPoolError::Network(e.to_string()))?)
    }
}
